cpp/ql/test/query-tests/Security/CWE/CWE-022/semmle/tests/TaintedPath.expected

```diff
@@ -1,4 +1,36 @@
+| test.c:37:17:37:24 | scanf output argument | test.c:38:11:38:18 | Convert indirection |
+| test.c:43:17:43:24 | scanf output argument | test.c:44:11:44:18 | Convert indirection |
+| test.c:57:10:57:13 | argv indirection | test.c:57:10:57:16 | access to array indirection |
+| test.c:57:10:57:13 | argv indirection | test.c:57:10:57:16 | access to array indirection |
 nodes
+| test.c:9:23:9:26 | argv | semmle.label | argv |
+| test.c:9:23:9:26 | argv indirection | semmle.label | argv indirection |
+| test.c:9:23:9:26 | argv indirection | semmle.label | argv indirection |
+| test.c:17:11:17:18 | Convert indirection | semmle.label | Convert indirection |
+| test.c:31:22:31:25 | argv indirection | semmle.label | argv indirection |
+| test.c:31:22:31:25 | argv indirection | semmle.label | argv indirection |
+| test.c:32:11:32:18 | Convert indirection | semmle.label | Convert indirection |
+| test.c:37:17:37:24 | scanf output argument | semmle.label | scanf output argument |
+| test.c:38:11:38:18 | Convert indirection | semmle.label | Convert indirection |
+| test.c:43:17:43:24 | scanf output argument | semmle.label | scanf output argument |
+| test.c:44:11:44:18 | Convert indirection | semmle.label | Convert indirection |
+| test.c:57:10:57:13 | argv indirection | semmle.label | argv indirection |
+| test.c:57:10:57:13 | argv indirection | semmle.label | argv indirection |
+| test.c:57:10:57:16 | access to array indirection | semmle.label | access to array indirection |
 subpaths
 #select
+| test.c:17:11:17:18 | fileName | test.c:9:23:9:26 | argv | test.c:17:11:17:18 | Convert indirection | This argument to a file access function is derived from $@ and then passed to fopen(filename). | test.c:9:23:9:26 | argv | user input (a command-line argument) |
+| test.c:17:11:17:18 | fileName | test.c:9:23:9:26 | argv indirection | test.c:17:11:17:18 | Convert indirection | This argument to a file access function is derived from $@ and then passed to fopen(filename). | test.c:9:23:9:26 | argv indirection | user input (a command-line argument) |
+| test.c:17:11:17:18 | fileName | test.c:9:23:9:26 | argv indirection | test.c:17:11:17:18 | Convert indirection | This argument to a file access function is derived from $@ and then passed to fopen(filename). | test.c:9:23:9:26 | argv indirection | user input (a command-line argument) |
+| test.c:32:11:32:18 | fileName | test.c:31:22:31:25 | argv indirection | test.c:32:11:32:18 | Convert indirection | This argument to a file access function is derived from $@ and then passed to fopen(filename). | test.c:31:22:31:25 | argv indirection | user input (a command-line argument) |
+| test.c:32:11:32:18 | fileName | test.c:31:22:31:25 | argv indirection | test.c:32:11:32:18 | Convert indirection | This argument to a file access function is derived from $@ and then passed to fopen(filename). | test.c:31:22:31:25 | argv indirection | user input (a command-line argument) |
```

Comment:
I believe I understand why we now get two results here, namely, there are several levels of indirection. I do wonder if this might be confusing for users when presented with these results.

Reply:
That's a good point. Initially, I included a number of `*`s in the `toString` to denote the number of indirections. But I decided against this since I was afraid of this involving too many concatenations at runtime.

We could try to undo this change, and generate `*argv` and `**argv` strings instead of assigning them both to `argv indirection`, and then see the impact on DCA or MRVA.

Alternatively, we could change the sink from `.asIndirectArgument()` to `.asIndirectArgument(1)` to only report the `*argv` case?

Finally, to get that ^ behavior by default, we could change the implementation of `Node.asIndirectArgument()` to be `this.asIndirectArgument(1)` instead of `this.asIndirectArgument(_)`.

Reply:
Should we just look at this as part of the tuning we want to do anyway?

Reply:
Yeah, I think that's a great idea.